### PR TITLE
Introduce `CompositeSink`

### PR DESCRIPTION
**Pull request by olivergrabinski**
Introduces the `CompositeSink`

* Previously the `QueryGraph`, the transform pipe, and the sink were separate pipes merged together.
* Now the `CompositeSink` is a single pipe that encompasses these operations.
* The goal is to make it easier to now introduce a second version of the `CompositeSink`

### CODE REVIEW

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeSink.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    override val maxWindow: FiniteDuration
+) extends Sink {
+
+  override type In = GraphResource
+  override def inType: Typeable[GraphResource] = Typeable[GraphResource]
+
+  private def query: Elem[GraphResource] => Task[Elem[GraphResource]] =
+    elem => queryGraph(elem)
+
+  private def liftedTransform: Elem[GraphResource] => Task[Elem[SinkFormat]] =
```
Comment (imsdu) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeSink.scala:43]: Has `query` and `transform` are two tasks, it should be possible to combine them this way no ?
`elem.evalMap(queryTransform)` so we would not have to do this

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/QueryGraph.scala [hunk 2]
```diff
@@ -43,7 +34,7 @@ final class QueryGraph(client: BlazegraphClient, namespace: String, query: Sparq
         Some(g.replaceRootNode(id))
       }
 
-  override def apply(element: SuccessElem[GraphResource]): Task[Elem[GraphResource]] =
+  def apply(element: Elem[GraphResource]): Task[Elem[GraphResource]] =
```
Comment (imsdu) [inline delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/QueryGraph.scala:37]: We can get rid of the elem here as it is not a pipe anymore

**Review by imsdu — CHANGES_REQUESTED** [review]

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeSink.scala (hunk 1 above)
Comment (olivergrabinski) [reply delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeSink.scala]: I made some changes, hopefully it's simpler now.

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/QueryGraph.scala (hunk 2 above)
Comment (olivergrabinski) [reply delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/QueryGraph.scala]: Removed

File: delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/GraphResourceToNTriples.scala [hunk 3]
```diff
@@ -19,16 +19,17 @@ object GraphResourceToNTriples extends Pipe {
...
-        err => element.failed(err),
-        {
-          case ntriples if ntriples.isEmpty => element.dropped
-          case ntriples                     => element.success(ntriples)
-        }
-      )
+    element
+      .evalMap(graphToNTriples)
+      .map {
+        case ntriples: SuccessElem[NTriples] if ntriples.value.isEmpty => element.dropped
```
Comment (imsdu) [inline delta/plugins/blazegraph/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/blazegraph/indexing/GraphResourceToNTriples.scala:31]: This part should also be in `graphToNTriples` to have a consistent behaviour when using the pipe or not.

File: delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/GraphResourceToDocument.scala [hunk 4]
```diff
@@ -31,22 +31,30 @@ final class GraphResourceToDocument(context: ContextValue, includeContext: Boole
...
         .map(json => injectId(json, id))
-        .map(json => if (json.isEmpty()) element.dropped else element.success(json))
     }
   }
 
+  override def apply(element: SuccessElem[GraphResource]): Task[Elem[Json]] =
+    element
+      .evalMap(graphToDocument)
+      .map {
+        case json: SuccessElem[Json] if json.value.isEmpty() => element.dropped
```
Comment (imsdu) [inline delta/plugins/elasticsearch/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/elasticsearch/indexing/GraphResourceToDocument.scala:54]: Same as the other case, it needs to be included in the previous method

**Review by imsdu — CHANGES_REQUESTED** [review]

**Review by imsdu — APPROVED** [review]

**Pull request merged**